### PR TITLE
#73: security portion of data slice support

### PR DESCRIPTION
**Pull request by brent-hartwig**
While this PR does not include the full implementation of data slices (in support of unit portals) and the data slice implementation needs to be vetted more rigorously, I'm comfortable with these changes reaching production as --for the most part-- we're just changing security roles and how document permissions are set.  The implementation leaves open the means to set document permissions as we have to date.  The new approach is dependent on the dataset including the `admin.sources` property (but the tenant owner will always have access to all of the documents).  Documentation will be ahead of its time and would have to be modified should data slices not come to full fruition. 

1. **Role changes:**
    * Introduced `base-reader` and `base-endpoint-consumer`.  The associated tenant and unit roles are to inherit these roles.
    * Reader roles no longer have the `rest-reader` role, rendering the use of reader roles to document permissions alone (and to be inherited by other roles).
    * Defined reader and endpoint consumer roles for the YCBA, YPM, and YUAG units.
    * **Disruptive:** Added the tenant name (`%%mlAppName%%`) to all unit role names.  This will impose one of the following on the LUX by Unit tenant.
        * Option 1: deploy the latest, reload the database, and update the unit service accounts to use the renamed roles.
        * Option 2: go ahead and deploy the latest but do not delete document permission roles or update unit service accounts until there is another reason to reload the database.  At that point, switch to Option 1.
    * **Disruptive:** For any other tenants/environments, when this implementation is rolled out *and* you wish to permission content at the unit level using the new approach, the database must be reloaded (or just initially loaded).
    * The following roles become obsolete once the content's document permissions account for this PR's role changes.
        * `%%mlAppName%%-ml-admin`
        * `%%mlAppName%%-admin`
        * `%%mlAppName%%-nobody`
        * `[unit]-*` (`%%mlAppName%%` was added to the unit role names)
2. Introduction of /src/main/ml-modules/root/documentTransforms.sjs, which is to be employed when loading content via MLCP or /v1/documents.  It is responsible for using the `admin.sources` property to set document permissions.  Gradle tasks updated.  MLCP example updated.  Discussed in lux-backend-security-and-software.md and lux-backend-deployment.md.
4. **Property changes:** changed `tenantEndpointConsumerPassword` and `tenantDeployerPassword` to `endpointConsumerPassword` and `deployerPassword`.  This only impacts unsecured deployments (i.e., local developer environments).
5. For unsecured deployments, defined YCBA, YPM, and YUAG endpoint consumer user accounts.
6. /scripts/checkTripleAndDocVisibilityByUser.js: find out which users can see which triples and documents.
7. /scripts/findHighlyRelatedRecords.js: helper script to find docs related to more records than other docs.
8. /scripts/getDocumentSampling.js: get a sampling of document URIs by one or more collections, plus additional constraints (optional).
9. Updated and added to the documentation.  
10. Related to #212, added a try/catch to highlight the likely reason `copyRestApiOptions` failed.

This PR makes the [73-research-data-slices branch](https://github.com/project-lux/lux-marklogic/tree/73-research-data-slices) obsolete.

There is outstanding work for #73.  See [this #73 comment](https://github.com/project-lux/lux-marklogic/issues/73#issuecomment-2195666094) for a partial list.

### CODE REVIEW

**Comment by brent-hartwig** [comment]
@clarkepeterf, this is another involved PR that we'd benefit from merging into the release branch to reduce the chance of merge conflicts and/or having to rebase.  Thank you.

File: build.gradle [hunk 1]
```diff
@@ -209,8 +209,8 @@ task importDataFull(type: com.marklogic.gradle.task.MlcpTask) {
   input_file_type = "delimited_json"
   input_compressed = project.property("importDataFileIsCompressed")
   input_compression_codec = "gzip"
-  output_permissions = project.property("mlDataPermissions")
```
Comment (clarkepeterf) [inline build.gradle:212]: if mlDataPermissions is no longer used, can it be removed from the properties files? Or is it used elsewhere implicitly by mlGradle? It is also placed into appConstants.mjs as a token replacement, but then the constant is never used.

Comment (brent-hartwig) [reply build.gradle:212]: Good catch.  It does look like it is becoming obsolete.  In this PR, I leave open the possibility of which way to permission content: see no. 1 in the [Tenant Limitations and Warnings](https://github.com/project-lux/lux-marklogic/blob/updated-data-slices/docs/lux-backend-deployment.md#tenant-limitations-and-warnings).  I'm good with removing now or later (if we can remember).

Comment (clarkepeterf) [reply build.gradle:212]: Gotcha. I'm ok with leaving it for now. If we decide to lean fully into permission setting via documentTransforms.sjs then let's try to remember to clean it up

**Pull request merged**